### PR TITLE
LSIF: Return null from GitBlob's LSIF resolver when no dump exists

### PR DESCRIPTION
**Pull request by efritz**
Ensure that GitBlob's lsif resolver returns null if there is no dump to answer LSIF queries.

### CODE REVIEW

File: cmd/frontend/graphqlbackend/schema.graphql [hunk 1]
```diff
@@ -2658,6 +2658,14 @@ type GitBlob implements TreeEntry & File2 {
 # A wrapper object around LSIF query methods for a particular path-at-revision. When this node is
 # null, no LSIF data is available for containing git blob.
 type LSIFQueryResolver {
+    # (experimental) The LSIF API may change substantially in the near future as we
+    # continue to adjust it for our use cases. Changes will not be documented in the
+    # CHANGELOG during this time.
+    # The commit that is being used to power intelligence. This may be distinct from
```
Comment (chrismwendt) [inline cmd/frontend/graphqlbackend/schema.graphql]:
```suggestion
    # The commit that is being used to power code intelligence. This may be distinct from
```

File: cmd/frontend/graphqlbackend/codeintel.go [hunk 2]
```diff
@@ -97,6 +98,7 @@ type LSIFQueryArgs struct {
 	RepoName string
 	Commit   GitObjectID
 	Path     string
+	DumpID   int64
```
Comment (chrismwendt) [inline cmd/frontend/graphqlbackend/codeintel.go]: Couple questions:

- When would a client specify this `DumpID`?
- Should this be added to the GraphQL schema? I don't see it there.

Comment (efritz) [reply cmd/frontend/graphqlbackend/codeintel.go]: No, this is just an implementation detail that's required by the lsif-server but hidden (purposefully) from the GraphQL client. This essentially does the same work as the closest commit dump ID that we restructured the RFC for.

Essentially, we need to hit the /exists endpoint to determine if a dump exists. Instead of doing the same dump lookup twice for a subsequent /definitions (or refs, hover) query, we can just pass the result to the subsequent request and save the closest commit query.

This isn't a strictly necessary change, but it doesn't affect the GraphQL API surface, not supplying it (accidentally or due to technical reasons) will not break functionality, and it will save some non-trivial amount (I suspect) of time spent in the database.

**Comment by efritz** [comment]
> Am I missing something, or does this PR not match the title+description?

Updated the title. It was pretty vague.

File: cmd/frontend/graphqlbackend/codeintel.go (hunk 2 above)
Comment (chrismwendt) [reply cmd/frontend/graphqlbackend/codeintel.go]: Huh, I wonder how important it is for a GraphQL implementation to mirror its schema, and whether or not saving a closest commit query is worth it. Did you determine that the closest commit query is a perf concern that's worth addressing? It's a factor of 2, right?

Comment (efritz) [reply cmd/frontend/graphqlbackend/codeintel.go]: My bigger concern is having race conditions from a single client returning an error rather than mixed results.

If you hit /exists and get a bool, you'll try to hit both /definitions and /references (for example) and will re-find the closest commit for each one. One query may load the dump and the other may load a different dump due to an upload being processed between the queries.

Tracking which dump ID we have for the top-level resolver will allow us some stability in the dump that is used for all of the queries for the same git blob for _that_ GraphQL query.

Reducing work is just a bonus on top of that. Instead of performing the same query up to four times, we only perform it once.

Comment (chrismwendt) [reply cmd/frontend/graphqlbackend/codeintel.go]: Ah, this makes sense for consistency :+1:

I was mistaken earlier when I thought the GraphQL implementation didn't match the schema. The `DumpID` isn't part of the GraphQL resolver method signature, it's a field on the resolver and a field on an intermediate struct.

File: cmd/frontend/graphqlbackend/schema.graphql [hunk 3]
```diff
@@ -2658,6 +2658,14 @@ type GitBlob implements TreeEntry & File2 {
 # A wrapper object around LSIF query methods for a particular path-at-revision. When this node is
 # null, no LSIF data is available for containing git blob.
 type LSIFQueryResolver {
+    # (experimental) The LSIF API may change substantially in the near future as we
+    # continue to adjust it for our use cases. Changes will not be documented in the
+    # CHANGELOG during this time.
+    # The commit that is being used to power code intelligence. This may be distinct from
+    # the commit of the git blob from which this query resolver came when there is no
+    # LSIF data available for that commit.
+    commit: GitCommit!
```
Comment (chrismwendt) [inline cmd/frontend/graphqlbackend/schema.graphql]: This is useful for the UI indicator, right? If I'm not mistaken, it doesn't appear to be related to the goal of this PR.

File: lsif/src/server/backend/backend.ts [hunk 4]
```diff
@@ -160,13 +160,19 @@ export class Backend {
...
+        repository: string,
+        commit: string,
+        path: string,
+        dumpId?: number,
+        ctx: TracingContext = {}
+    ): Promise<xrepoModels.LsifDump | undefined> {
+        const closestDatabaseAndDump = await this.loadClosestDatabase(repository, commit, path, dumpId, ctx)
         if (!closestDatabaseAndDump) {
-            return false
+            return undefined
```
Comment (chrismwendt) [inline lsif/src/server/backend/backend.ts]: I believe this will not require changes to basic-code-intel because it's truthy/falsy in the same way https://github.com/sourcegraph/sourcegraph-basic-code-intel/blob/e19642f025c2a9ab6ccf67bdad3a7959985fa2a1/package/src/lsif.ts#L118-L119

Does that seem to be the case to you, too?

Comment (efritz) [reply lsif/src/server/backend/backend.ts]: https://github.com/sourcegraph/sourcegraph/blob/5176d0b68d9b830b4e339fd7b29cd6fb178957fa/lsif/src/server/routes/lsif.ts#L281

My goal was to not destroy any existing endpoint (yet).

File: cmd/frontend/graphqlbackend/schema.graphql (hunk 3 above)
Comment (efritz) [reply cmd/frontend/graphqlbackend/schema.graphql]: Will put in another PR.

**Comment by unknwon** [comment]
Sorry about coming late, files owned by core services LGTM.
